### PR TITLE
RetinaNet MLPerf

### PR DESCRIPTION
**Pull request by francislata**
# Overview
This PR adds the MLPerf training script for the RetinaNet model.

# How to run the training script
Run the following command to install dependencies:
`pip install tqdm numpy wandb pycocotools boto3 pandas torch torchvision`

To download the `openimages` dataset, run the following:
`BASE_DIR=/raid/datasets/openimages/ python3 extra/datasets/openimages.py`
 
Once the dataset has been downloaded, use the following train command:
`WANDB=1 DEFAULT_FLOAT=HALF BEAM_PADTO=0 BEAM_UOPS_MAX=1500 BEAM_UPCAST_MAX=64 BEAM_MIN_PROGRESS=5 TRAIN_BEAM=4 IGNORE_BEAM_CACHE=1 IGNORE_JIT_FIRST_BEAM=1 MODEL=retinanet BASE_DIR=/raid/datasets/openimages GPUS=6 python3 examples/mlperf/model_train.py`

This will use a batch size of 96. Here is a [complete run](https://wandb.ai/flata/MLPerf-RetinaNet/runs/6je5509t/overview) that finished at around 18h2m with a metric of `0.3454`.

# Tasks
Since this is still in draft mode, here are some `TODO`s that I want to get done before it is ready for review:
- [x] Add validation test run in training script
- [x] Add W&B integration
- [x] Perform a full training run of model
- [x] Update newly-added typing annotations

# PR breakdown
Once all `TODO`s are done and correctness is verified, here's the breakdown of independent PRs that I can put up:
- [x] #9442
- [x] #9536
- [x] #9773
- [x] #9747 
- [x] #9822

### CODE REVIEW

**Comment by francislata** [comment]
Here are the W&B correctness runs in `float32`:
- [tinygreen](https://wandb.ai/flata/MLPerf-RetinaNet/runs/pnrnj5gw?nw=nwuserflata)
- [tinyred](https://wandb.ai/flata/MLPerf-RetinaNet/runs/yzfyh8x0?nw=nwuserflata)

File: test/external/mlperf_retinanet/boxes.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+# https://github.com/mlcommons/training/blob/cdd928d4596c142c15a7d86b2eeadbac718c8da2/single_stage_detector/ssd/model/boxes.py
+
+import torch
+from torch import Tensor
+from typing import Tuple
+
+
+def _upcast(t: Tensor) -> Tensor:
+    # Protects from numerical overflows in multiplications by upcasting to the equivalent higher type
```
Comment (chenyuxyz) [inline test/external/mlperf_retinanet/boxes.py:9]: indentation

**Comment by francislata** [comment]
For reference, here is [an float16 run](https://wandb.ai/flata/MLPerf-RetinaNet/runs/52okpz5k/overview) that finished around `23h44m` with a final metric of `0.34146` by using the following command:
`DEBUG=2 LR=0.000085 EPOCHS=8 WANDB=1 DEFAULT_FLOAT=HALF BEAM_UOPS_MAX=1500 BEAM_UPCAST_MAX=64 BEAM_PADTO=0 BEAM_MIN_PROGRESS=5 TRAIN_BEAM=4 IGNORE_BEAM_CACHE=1 IGNORE_JIT_FIRST_BEAM=1 LR_WARMUP_EPOCHS=0 MODEL=retinanet BASE_DIR=/raid/datasets/openimages BS=114 GPUS=6 python3 examples/mlperf/model_train.py`

File: examples/mlperf/model_train.py [hunk 2]
```diff
@@ -343,8 +343,269 @@ def data_get(it):
...
+  coco_val = COCOeval(cocoGt=val_dataset, iouType="bbox")
+
+  # ** lr scheduler **
+  config["steps_in_train_epoch"] = steps_in_train_epoch = round_up(len(train_dataset.imgs.keys()), BS) // BS
+  config["steps_in_val_epoch"] = steps_in_val_epoch = (round_up(len(val_dataset.imgs.keys()), BS) // BS)
+  start_iter = start_epoch * steps_in_train_epoch
+
+  # ** resume from checkpointing **
+  if (ckpt:=getenv("RESUME", "")):
+    # TODO: Address resuming from checkpoint
```
Comment (francislata) [inline examples/mlperf/model_train.py:454]: @chenyuxyz let me know if this is a blocker for this PR. i can comeback to this on a separate PR. currently, the helper function needs an LR scheduler to get some metadata information like the current epoch, etc. so i'll have to modify the helper function to optionally support it.

alternatively, i can remove the saving code from this PR and add it back + make the helper function changes.

File: examples/mlperf/initializers.py [hunk 3]
```diff
@@ -66,3 +66,62 @@ def __call__(self, x:Tensor):
     xn = x.cast(dtypes.float32).layernorm(eps=self.eps, axis=self.axis).cast(x.dtype)
     if not self.elementwise_affine: return xn
     return (xn * self.weight.cast(dtypes.default_float) + self.bias.cast(dtypes.default_float))
+
+class FrozenBatchNorm2d(nn.BatchNorm2d):
```
Comment (chenyuxyz) [inline examples/mlperf/initializers.py:70]: lack of a better way to manage these, can you add `Retinanet` to these class names to not confused with the original ones?

File: examples/mlperf/model_train.py [hunk 4]
```diff
@@ -343,8 +343,269 @@ def data_get(it):
...
+  @TinyJit
+  def _train_step(model, optim, loss_scaler, x, **kwargs):
+    with Context(BEAM=TRAIN_BEAM):
+      optim.zero_grad()
+
+      losses = model(normalize(x, GPUS), **kwargs)
+      loss = sum([l for l in losses.values()])
+
+      (loss * loss_scaler).backward()
+      for t in optim.params: t.grad = t.grad.contiguous() / loss_scaler
```
Comment (chenyuxyz) [inline examples/mlperf/model_train.py:393]: not a blocker, probably fine to drop the .contiguous

File: examples/mlperf/model_train.py (hunk 2 above)
Comment (chenyuxyz) [reply examples/mlperf/model_train.py:454]: if it's not working just remove all these, it's not a requirement to have checkpoint resume

File: examples/mlperf/model_train.py [hunk 5]
```diff
@@ -343,8 +343,269 @@ def data_get(it):
...
+    # load_training_state(model, optim, lr_scheduler, safe_load(ckpt))
+    # start_epoch = int(lr_scheduler.epoch_counter.item() / steps_in_train_epoch)
+    print(f"resuming from {ckpt} at epoch {start_epoch}")
+
+  # ** initialize wandb **
+  if (WANDB:=getenv("WANDB")):
+    import wandb
+
+    wandb_args = {"project": "MLPerf-RetinaNet"}
+    if (wandb_id:=getenv("WANDB_RESUME", "")):
```
Comment (chenyuxyz) [inline examples/mlperf/model_train.py:464]: does WANDB_RESUME work, if no just remove it. keep the pr as small as possible

File: examples/mlperf/model_train.py [hunk 6]
```diff
@@ -343,8 +343,269 @@ def data_get(it):
...
+      coco_val.summarize()
+
+      val_metric = coco_val.stats[0]
+
+      tqdm.write(f"eval time: {total_fw_time:.2f}, eval metric: {val_metric:.4f}")
+
+      if WANDB:
+        wandb.log({"eval/forward_time": total_fw_time, "eval/metric": val_metric, "epoch": e + 1})
+
+      if getenv("CKPT", 1):
```
Comment (chenyuxyz) [inline examples/mlperf/model_train.py:594]: remove if not used

File: extra/models/retinanet.py [hunk 7]
```diff
@@ -30,26 +35,8 @@ def decode_bbox(offsets, anchors):
...
-    shifts_x, shifts_y = np.meshgrid(np.arange(gs[1]) * stride_w, np.arange(gs[0]) * stride_h)
-    shifts_x = shifts_x.reshape(-1)
-    shifts_y = shifts_y.reshape(-1)
-    shifts = np.stack([shifts_x, shifts_y, shifts_x, shifts_y], axis=1, dtype=np.float32)
-    anchors.append((shifts[:, None] + base_anchors[None, :]).reshape(-1, 4))
-  return anchors
-
 class RetinaNet:
-  def __init__(self, backbone: ResNet, num_classes=264, num_anchors=9, scales=None, aspect_ratios=None):
+  def __init__(self, backbone:ResNet, num_classes:int=264, num_anchors:int=9, scales:list[int]|None=None, aspect_ratios:list[float]|None=None):
```
Comment (chenyuxyz) [inline extra/models/retinanet.py:39]: make type annotation its own pr, with other **kwargs changes too

Comment (francislata) [reply extra/models/retinanet.py:39]: i've added it, plus all of the loss function changes, to #9822.
